crypto/s2n_aead_cipher_chacha20_poly1305.c

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright 2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <openssl/evp.h>
+
+#include "crypto/s2n_cipher.h"
+#include "crypto/s2n_openssl.h"
+
+#include "tls/s2n_crypto.h"
+
+#include "utils/s2n_safety.h"
+#include "utils/s2n_blob.h"
+
+static const EVP_CIPHER *s2n_evp_chacha20_poly1305(void)
+{
+    /* EVP for ChaCha20-Poly1305 added in Openssl 1.1.0. See:
+     * https://www.openssl.org/news/cl110.txt . No support in LibreSSL.
+     */
+    #if S2N_OPENSSL_VERSION_AT_LEAST(1,1,0) && !defined LIBRESSL_VERSION_NUMBER
+        return EVP_chacha20_poly1305();
+    #else
+        return NULL;
+    #endif
+}
+
+static uint8_t s2n_aead_chacha20_poly1305_available(void)
+{
+    return (s2n_evp_chacha20_poly1305() ? 1 : 0);
+}
+
+static int s2n_aead_chacha20_poly1305_encrypt(struct s2n_session_key *key, struct s2n_blob *iv, struct s2n_blob *aad, struct s2n_blob *in, struct s2n_blob *out)
+{
```

Comment:
I was really hoping to avoid having the macros in-line with the core code. It looks like the LibreSSL problems are going to force us to use this approach.